app/controllers/users_controller.rb

```diff
@@ -119,4 +99,20 @@ def user_from_params
         user.password = password
       end
     end
+
+    def update_protected_params
+      if @user.authenticated?(user_params[:password])
```

Comment:
we don't need three cases here but two cases: 
1. if the user is authenticated - in which case the data should be updated
2. if the user has not authenticated their password, in case we should display an error
the only difference between the empty password and the wrong password is the message being displayed in the flash message. for DRY I would go for a conditional to define the message and render :edit if the user has not authenticated